[LinalgExt] Initial support for aten::flex_attention and rewrite to linalg_ext.attention

keshavvinayak01 opened the pull request:
This PR adds comprehensive support for converting PyTorch's `torch::hop_flex_attention` operation to IREE's `linalg_ext.attention` operation. The implementation includes proper handling of score modification functions, mask functions, and correct softmax computation using base-e exponential (exp) instead of base-2 exponential (exp2).

## Changes

- Added `FlexAttentionOpConversion` pattern to convert `torch.hop_flex_attention` to `linalg_ext.attention`
- Modified `AttentionOp` to pass iteration indices as block arguments to the region
- Enables score modification functions to access batch, head, query sequence, and key sequence indices
- Added lit tests for both LSE (log-sum-exp) and non-LSE cases
- Simplified region handling by using block arguments instead

## Fixes:
- PyTorch's flex attention already supplies the correct scale for base-e softmax. This commit fixes the computation to use `exp` instead of `exp2`:
- The use_exp2 flag is mostly unused in dialect conversions and passes, I presume it's used as a KernelOption. The changes here will not modify the default behavior.

## Testing:
- I ran the entire flex_attention_hop implementation with randomised input tensors, (Also see [torch-mlir](https://github.com/llvm/torch-mlir/pull/4366)) through `aot.export` and compared against eager mode, and I noticed no accuracy losses (On CPU)
- Command: `iree-compile --iree-stream-partitioning-favor=min-peak-memory --iree-hal-target-device=local --iree-hal-local-target-device-backends=llvm-cpu --iree-llvmcpu-target-triple=x86_64-pc-linux-elf --iree-llvmcpu-debug-symbols=false ../torch-mlir/exported_ir.mlir -o ./flex_attention_cpu.vmfb`

keshavvinayak01 committed:
Initial support for aten::flex_attention and rewrite to linalgext.online_attention

Signed-off-by: Keshav Vinayak Jha <keshavvinayakjha@gmail.com>

keshavvinayak01 committed:
Pass iteration indices as block arguments to OnlineAttentionOp region

Instead of requiring IREE::LinalgExt::IndexOp inside the region body,
pass iteration indices (b, h, m, n) directly as block arguments. This
simplifies lowering passes.

Changes:
- Update OnlineAttentionOp/AttentionOp verifiers to accept 1 or 5 block args
- Modify applyPostQKMatmulElementwise to provide indices as block arguments
- Update FlexAttentionOp lowering to declare 5 block arguments
- Document region signature in LinalgExtOps.td

The region now receives: (score: f32, b: index, h: index, m: index, n: index)
instead of requiring IndexOp operations to query indices.

Signed-off-by: Keshav Vinayak Jha <keshavvinayakjha@gmail.com>

keshavvinayak01 committed:
Initialized reduction vals; Corrected func::call op init

Signed-off-by: Keshav Vinayak Jha <keshavvinayakjha@gmail.com>

keshavvinayak01 committed:
Added lit test (lse, nolse)

Signed-off-by: Keshav Vinayak Jha <keshavvinayakjha@gmail.com>

keshavvinayak01 committed:
Removed previous support added for linalgext::index op to nest in attention region

Signed-off-by: Keshav Vinayak Jha <keshavvinayakjha@gmail.com>

keshavvinayak01 committed:
Formatting

Signed-off-by: Keshav Vinayak Jha <keshavvinayakjha@gmail.com>

keshavvinayak01 committed:
Added changes to use IREE::AttentionOp; support reqiured for multi-dimension access

Signed-off-by: Keshav Vinayak Jha <keshavvinayakjha@gmail.com>

keshavvinayak01 committed:
Removed unused comment

Signed-off-by: Keshav Vinayak Jha <keshavvinayakjha@gmail.com>

keshavvinayak01 committed:
Added changes for correct computations:
-  PyTorch’s attention already supplies the correct scale for the base‑e softmax. All the following files are changed to support exp computation instead of exp2.
1. ConvertTorchUnstructuredToLinalgExt: FlexAttentionOpConversion pattern passes use_exp2 = false, which can be used correctly in decomposition.
2. AggregatedOpInterfaceImpl: accepts the use_exp2 flag as an attribute for decomposition and calls computeSubAndExp accordingly.
3. LinalgExtOps.td: Added getUseExp2AttrStr() to both online_attention and attention.
4. ReshapeFusion.cpp: createCollapsedOP was recreating the attention op and stripping all attributes before StripCompilationPass. Change was necessary to support correct decomposition.

Signed-off-by: Keshav Vinayak Jha <keshavvinayakjha@gmail.com>

keshavvinayak01 committed:
.create -> OpTy::create

Signed-off-by: Keshav Vinayak Jha <keshavvinayakjha@gmail.com>

keshavvinayak01 committed:
Removed gqa argument

Signed-off-by: Keshav Vinayak Jha <keshavvinayakjha@gmail.com>

hanhanW commented:
I can review after you fix all build failures. It looks not ready to me when CI is so red.

hanhanW commented on compiler/plugins/input/Torch/InputConversion/ConvertTorchUnstructuredToLinalgExt.cpp:
```diff
@@ -163,6 +171,349 @@ struct FftRfftOpConversion
+    // Query shape: [B, H, M, E]
+    if (queryShape.size() != 4) {
+      return rewriter.notifyMatchFailure(op, "expected 4D query tensor");
+    }
```
ditto

hanhanW commented on compiler/plugins/input/Torch/InputConversion/ConvertTorchUnstructuredToLinalgExt.cpp:
```diff
@@ -163,6 +171,349 @@ struct FftRfftOpConversion
+    if (!matchPattern(scaleValue,
+                      torch::Torch::m_TorchConstantFloat(&scaleVal)))
+      // Fallback to default scale
+      scaleVal = 1.0 / std::sqrt(static_cast<double>(headDim));
```
IREE style nit: wrap braces. Please check other code as well.

note: you also need braces even you are following LLVM style: https://llvm.org/docs/CodingStandards.html#don-t-use-braces-on-simple-single-statement-bodies-of-if-else-loop-statements

hanhanW commented on compiler/plugins/input/Torch/InputConversion/ConvertTorchUnstructuredToLinalgExt.cpp:
```diff
@@ -163,6 +171,349 @@ struct FftRfftOpConversion
+      Value negInf = arith::ConstantOp::create(
+          rewriter, loc, floatType,
+          rewriter.getFloatAttr(floatType,
+                                -std::numeric_limits<double>::infinity()));
```
For zero, you can use `arith::ConstantFloatOp::create(..., floatType, 0)`.

For negInf, you should try this method: https://github.com/llvm/llvm-project/blob/3a2de951a40877e3d17aa87abb35565da1040612/llvm/include/llvm/ADT/APFloat.h#L1080-L1084

`double` != floatType. The precision can mismatch.

hanhanW commented on compiler/plugins/input/Torch/InputConversion/test/unstructured_linalg_ext.mlir:
```diff
@@ -99,3 +99,150 @@ func.func @fft_rfft.last(%arg0: !torch.vtensor<[3,8,16],f32>) -> !torch.vtensor<
+func.func private @sdpa_mask0(%arg0: !torch.vtensor<[],si32>, %arg1: !torch.vtensor<[],si32>, %arg2: !torch.vtensor<[],si32>, %arg3: !torch.vtensor<[],si32>) -> !torch.vtensor<[],i1> {
+  %0 = torch.aten.ge.Tensor %arg2, %arg3 : !torch.vtensor<[],si32>, !torch.vtensor<[],si32> -> !torch.vtensor<[],i1>
+  return %0 : !torch.vtensor<[],i1>
+}
```
This is op level lowering test. Do you really need all of them? Can you simplify it to a single attention op?

keshavvinayak01 replied:
I could rewrite a simpler score function, but yes, I need an op lowering test that demonstrates we can properly rewrite functional references.

hanhanW commented on compiler/plugins/input/Torch/InputConversion/ConvertTorchUnstructuredToLinalgExt.cpp:
```diff
@@ -171,13 +522,15 @@ class ConvertTorchUnstructuredToLinalgExtPass final
     RewritePatternSet patterns(context);
 
     patterns.add<FftRfftOpConversion>(context);
+    patterns.add<FlexAttentionOpConversion>(context);
```
Merge them.

keshavvinayak01 commented on compiler/plugins/input/Torch/InputConversion/ConvertTorchUnstructuredToLinalgExt.cpp:
```diff
@@ -163,6 +171,349 @@ struct FftRfftOpConversion
+    if (!queryType.hasSizes() || !keyType.hasSizes() || !valueType.hasSizes()) {
+      return rewriter.notifyMatchFailure(op,
+                                         "expected tensors with known sizes");
+    }
```
Yeah, it's a small check to verify the input types has sizes, required by the rewrite. I'll remove the comments.

keshavvinayak01 commented on compiler/plugins/input/Torch/InputConversion/ConvertTorchUnstructuredToLinalgExt.cpp:
```diff
@@ -163,6 +171,349 @@ struct FftRfftOpConversion
+    decompositionConfigAttrs.push_back(
+        rewriter.getNamedAttr("use_exp2", rewriter.getBoolAttr(false)));
+    DictionaryAttr decompositionConfig =
+        rewriter.getDictionaryAttr(decompositionConfigAttrs);
```
Thanks for pointing this out, this is a massive assumption, and the computation _might_ vary backend-to-backend. Maybe I can pass a flag in the `iree-linalg-ext-decompose-attention` pass? I verified we need to use `exp2` there, at least on the CPU for correct computations. While testing for GPU, it'd give me the liberty to turn that flag or not, assuming that pass is even part of the pipeline (Most likely not).

hanhanW replied:
> Maybe I can pass a flag in the iree-linalg-ext-decompose-attention pass?

Starting it with this is better than the setting it in the lowering.

> I verified we need to use exp2 there, at least on the CPU for correct computations.

I haven't reviewed the details yet, but it sounds like numeric difference from fp re-association. The golden values from the framework might be off, or we need to adjust tolerance for the op test.

keshavvinayak01 committed:
- Removed redundant/debug comments
- General formatting
- Removed hardcoded exp2 decomposition attribute

Signed-off-by: Keshav Vinayak Jha <keshavvinayakjha@gmail.com>

IanWood1 commented on compiler/src/iree/compiler/Dialect/LinalgExt/IR/LinalgExtOps.td:
```diff
@@ -798,6 +789,12 @@ def IREELinalgExt_AttentionOp : IREELinalgExt_Op<"attention",
+    Region:
+      The region body can receive either 1 or 5 block arguments:
+        - 1 argument (legacy): score (element type of output)
+        - 5 arguments: score, b (batch index), h (head index), m (query seq index), n (key/value seq index)
```
Is the idea that there will be a follow-up to remove support for the "legacy" 1 argument version?

keshavvinayak01 committed:
Added missing newline

Signed-off-by: Keshav Vinayak Jha <keshavvinayakjha@gmail.com>

keshavvinayak01 committed:
Addressed comments and changes:
1. ConvertTorchUnstructuredToLinalgExt.cpp: minor formatting stuff, and removed a redundant comment.
2. unstructured_linalg_ext.mlir: Modified lit test with smaller score function for simplicity; Added warning check incase lse output is expected.
3. AggregatedOpInterfaceImp.cpp: AttentionOp doesn't need the exp2 modification. Removed redundant comments.
4. LinalgExtOps.td: AttentionOp doesn't need getUseExp2AttrStr() attribute.
5. decompose_aggregate_op.mlir: Added another online_attention op variant that passes use_exp2=False in the decompositionConfig and CHECKs that math.exp2 ops are not created.
6. DecomposeAttention.cpp: Extract useExp2 bool from the pass option and attaches a new decompositionconfigattr with the relevant value.
7. Passes.td: iree-linalg-ext-decompose-attention pass now accepts useExp2 as a pass option (Defaults to true to preserve legacy behavior)
8. ReshapeFusion.cpp: I had forced the collapsedOp to again attach the decompositionConfig, don't need this change anymore.

Signed-off-by: Keshav Vinayak Jha <keshavvinayakjha@gmail.com>

keshavvinayak01 committed:
Formattingh

Signed-off-by: Keshav Vinayak Jha <keshavvinayakjha@gmail.com>

keshavvinayak01 committed:
Restored ReshapeFusion.cpp to original format

Signed-off-by: Keshav Vinayak Jha <keshavvinayakjha@gmail.com>

keshavvinayak01 committed:
Added 4 more lit tests covering with/without score_mod, mask_mod, and lse return

Signed-off-by: Keshav Vinayak Jha <keshavvinayakjha@gmail.com>

keshavvinayak01 committed:
Smaller matchAndRewrite calls to specific utility functions; easily readable

Signed-off-by: Keshav Vinayak Jha <keshavvinayakjha@gmail.com>

keshavvinayak01 committed:
Replaced magic numbers with meaningful consts

Signed-off-by: Keshav Vinayak Jha <keshavvinayakjha@gmail.com>

keshavvinayak01 commented:
@MaheshRavishankar Thanks for the review, I'll incorporate some of the suggested changes. 
As for the PR bifurcation, It's a bit difficult to isolate the lowering and the required changes. The lowering relies on block arguments being supported which is a change to the AttentionOp interface itself.
Could you be more specific about how you see this split being done?

keshavvinayak01 commented:
> > @MaheshRavishankar Thanks for the review, I'll incorporate some of the suggested changes. As for the PR bifurcation, It's a bit difficult to isolate the lowering and the required changes. The lowering relies on block arguments being supported which is a change to the AttentionOp interface itself. Could you be more specific about how you see this split being done?
> 
> Yeah, first you should make a change to the `AttentionOp` itself which will get the block arguments support. Then you can make the changes to the lowering? The part adding block arguments here needs to be flushed out a bit more. Do we really need the change to get block arguments in? One option is to not change block arguments but use `linalg_ext.index` (similar to `linalg.index). That might be easier to land without too many breakages.

Previously, that's how the implementation looked like, I added support for `linalg_ext.index` ops inside the attention region. That worked, but bloated up the IR, so I thought this was a simpler solution.

 In any case, It's easier to support index ops inside the attention region (With small modifications to the `::verify` method), so I'll switch back to it, because it's less likely to break things in the future.

Regardless, I'll have to add tiling support for these index ops in the attention region, which will be a follow up PR.

keshavvinayak01 commented:
I had discussed this with @Groverkss earlier, the goal was to add support for iree_linalg_ext.index in the attention op region to enable index-based logic (e.g., masking for RPE). At the time, he pointed out that using the `mask` operand in `AttentionOp` was a hack.

While most tiled ops don’t require index in the region, this op does. It relies on index-dependent semantics like relative position encoding, which can’t be expressed correctly without index after tiling.
